File: model_upgrade/dense.py
"""Dense lambda graphs: max-clique via max independent set on the complement."""

import random
import time

from model_upgrade.bitsets import (
    MAX_BITSET_NODES,
    mask_degree,
    mask_intersection,
    mask_vertices,
    neighbor_sets_to_masks,
)
from model_upgrade.graph_utils import core_and_degeneracy
from model_upgrade.validation import extend_to_maximal_clique, is_valid_maximum_clique

# Validator lambda graphs: complement search from ~70% edge density upward.
DENSE_DEGREE_RATIO = 0.70


def is_dense_graph(
    neighbor_sets: list[set[int]],
    threshold: float = DENSE_DEGREE_RATIO,
) -> bool:
    """True when the graph is a dense lambda instance (complement is sparse)."""
    n = len(neighbor_sets)
    if n <= 1:
        return False
    avg_degree = sum(len(neighbors) for neighbors in neighbor_sets) / n
    return avg_degree / (n - 1) >= threshold


def complement_neighbor_sets(neighbor_sets: list[set[int]]) -> list[set[int]]:
    """Non-neighbors in G (neighbors in the complement graph)."""
    n = len(neighbor_sets)
    universe = set(range(n))
    return [universe - {vertex} - neighbor_sets[vertex] for vertex in range(n)]


def _iter_bits(mask: int):
    while mask:
        low = mask & -mask
        yield low.bit_length() - 1
        mask &= ~low


def greedy_mis(
    comp_neighbors: list[set[int]],
    vertex_order: list[int],
) -> list[int]:
    """Greedy independent set on the complement graph."""
    mis: list[int] = []
    blocked: set[int] = set()
    for vertex in vertex_order:
        if vertex in blocked:
            continue
        mis.append(vertex)
        blocked.add(vertex)
        blocked.update(comp_neighbors[vertex])
    return mis


def greedy_mis_grasp(
    comp_neighbors: list[set[int]],
    rng: random.Random,
    start_vertex: int | None = None,
    rcl_size: int = 5,
    comp_masks: list[int] | None = None,
) -> list[int]:
    """GRASP independent-set construction (low complement-degree first)."""
    n = len(comp_neighbors)
    use_masks = comp_masks is not None and n <= MAX_BITSET_NODES

    mis: list[int] = []
    if start_vertex is not None:
        mis = [start_vertex]

    if use_masks:
        assert comp_masks is not None
        blocked_mask = 0
        for vertex in mis:
            blocked_mask |= (1 << vertex) | comp_masks[vertex]

        while True:
            available_mask = ((1 << n) - 1) & ~blocked_mask
            if not available_mask:
                break
            candidates = mask_vertices(available_mask)
            ranked = sorted(
                candidates,
                key=lambda v: mask_degree(comp_masks[v] & available_mask),
            )
            top = ranked[: min(rcl_size, len(ranked))]
            chosen = rng.choice(top)
            mis.append(chosen)
            blocked_mask |= (1 << chosen) | comp_masks[chosen]
        return mis

    blocked: set[int] = set()
    for vertex in mis:
        blocked.add(vertex)
        blocked.update(comp_neighbors[vertex])

    while True:
        available = set(range(n)) - blocked
        if not available:
            break
        ranked = sorted(
            available,
            key=lambda v: len(comp_neighbors[v] & available),
        )
        top = ranked[: min(rcl_size, len(ranked))]
        chosen = rng.choice(top)
        mis.append(chosen)
        blocked.add(chosen)
        blocked.update(comp_neighbors[chosen])
    return mis


def mis_random_restarts(
    comp_neighbors: list[set[int]],
    deadline: float,
    rng: random.Random,
    comp_masks: list[int] | None = None,
) -> list[int]:
    """Multi-start greedy / GRASP independent-set construction."""
    n = len(comp_neighbors)
    comp_degrees = [len(neighbors) for neighbors in comp_neighbors]
    best: list[int] = []

    static_orders = [
        sorted(range(n), key=lambda v: comp_degrees[v]),
        sorted(range(n), key=lambda v: comp_degrees[v], reverse=True),
        sorted(range(n), key=lambda v: (comp_degrees[v], v)),
        list(range(n)),
        list(reversed(range(n))),
    ]
    _, degeneracy = core_and_degeneracy(comp_neighbors)
    static_orders.append(list(degeneracy))
    static_orders.append(list(reversed(degeneracy)))

    for order in static_orders:
        if time.perf_counter() >= deadline:
            break
        candidate = greedy_mis(comp_neighbors, order)
        if len(candidate) > len(best):
            best = candidate

    sparse_starts = sorted(range(n), key=lambda v: comp_degrees[v])[: min(24, n)]
    for start in sparse_starts:
        if time.perf_counter() >= deadline:
            break
        candidate = greedy_mis_grasp(
            comp_neighbors,
            rng,
            start_vertex=start,
            comp_masks=comp_masks,
        )
        if len(candidate) > len(best):
            best = candidate

    while time.perf_counter() < deadline:
        if rng.random() < 0.55 and sparse_starts:
            start = rng.choice(sparse_starts)
            candidate = greedy_mis_grasp(
                comp_neighbors,
                rng,
                start_vertex=start,
                rcl_size=rng.randint(4, 8),
                comp_masks=comp_masks,
            )
        else:
            order = list(range(n))
            rng.shuffle(order)
            candidate = greedy_mis(comp_neighbors, order)
        if len(candidate) > len(best):
            best = candidate

    return best


def _mis_compatible(comp_neighbors: list[set[int]], mis_set: set[int], vertex: int) -> bool:
    return not (mis_set & comp_neighbors[vertex])


def mis_local_search(
    comp_neighbors: list[set[int]],
    mis: list[int],
    deadline: float,
    rng: random.Random,
) -> list[int]:
    """1-swap and (1,2)-add plateau search on the complement (clique in G)."""
    best = list(mis)
    best_set = set(best)
    n = len(comp_neighbors)

    while time.perf_counter() < deadline:
        if not best:
            break

        remove_vertices: list[int]
        if rng.random() < 0.2 and len(best) >= 2:
            ranked = sorted(best, key=lambda v: len(comp_neighbors[v]))
            remove_vertices = ranked[:2]
        elif rng.random() < 0.75:
            remove_vertices = [min(best, key=lambda v: len(comp_neighbors[v]))]
        else:
            remove_vertices = [rng.choice(best)]

        reduced = best_set - set(remove_vertices)
        candidates = [
            vertex
            for vertex in range(n)
            if vertex not in reduced and _mis_compatible(comp_neighbors, reduced, vertex)
        ]
        if not candidates:
            continue

        improved = False

        if len(remove_vertices) == 1:
            add_vertex = min(candidates, key=lambda v: len(comp_neighbors[v]))
            trial = sorted(reduced | {add_vertex})
            if len(trial) >= len(best):
                best = trial
                best_set = set(best)
                improved = True

            if not improved and len(best) >= 2 and rng.random() < 0.4:
                for i, u in enumerate(candidates):
                    for v in candidates[i + 1 :]:
                        if v not in comp_neighbors[u]:
                            pair_trial = sorted(reduced | {u, v})
                            if len(pair_trial) > len(best):
                                best = pair_trial
                                best_set = set(best)
                                improved = True
                                break
                    if improved:
                        break

        if not improved and len(remove_vertices) == 1:
            add_vertex = rng.choice(candidates)
            trial = sorted(reduced | {add_vertex})
            if len(trial) == len(best):
                best = trial
                best_set = set(best)

    return best


def mis_bitset_local_search(
    comp_masks: list[int],
    n: int,
    init_mis: list[int],
    deadline: float,
    rng: random.Random,
    penalties: list[int] | None = None,
) -> list[int]:
    """Bitset local search on sparse complement adjacency."""
    mis = list(init_mis)
    mis_mask = 0
    for vertex in mis:
        mis_mask |= 1 << vertex

    best = list(mis)
    best_size = len(mis)
    if penalties is None:
        penalties = [0] * n

    full_mask = (1 << n) - 1

    def available_mask() -> int:
        blocked = mis_mask
        for vertex in mis:
            blocked |= comp_masks[vertex]
        return full_mask & ~blocked

    avail = available_mask()
    check_counter = 0

    while True:
        check_counter += 1
        if check_counter >= 256:
            check_counter = 0
            if time.perf_counter() >= deadline:
                break

        if avail:
            if rng.random() < 0.65:
                chosen = -1
                chosen_score = None
                for vertex in _iter_bits(avail):
                    score = (comp_masks[vertex] & avail).bit_count() - penalties[vertex]
                    if chosen_score is None or score < chosen_score:
                        chosen_score = score
                        chosen = vertex
            else:
                chosen = rng.choice(list(_iter_bits(avail)))

            mis.append(chosen)
            mis_mask |= 1 << chosen
            avail = available_mask()

            if len(mis) > best_size:
                best = list(mis)
                best_size = len(mis)
            continue

        if not mis:
            break

        improved = False
        for _ in range(min(len(mis), 4)):
            pivot = rng.choice(mis)
            reduced_mask = mis_mask & ~(1 << pivot)
            mask = full_mask & ~reduced_mask
            for vertex in mis:
                if vertex != pivot:
                    mask &= ~comp_masks[vertex]
            mask &= ~reduced_mask

            pair = None
            for first in _iter_bits(mask):
                rest = mask & ~comp_masks[first] & ~(1 << first)
                if rest:
                    second = (rest & -rest).bit_length() - 1
                    pair = (first, second)
                    break

            if pair is not None:
                first, second = pair
                mis.remove(pivot)
                mis_mask &= ~(1 << pivot)
                mis.extend((first, second))
                mis_mask |= (1 << first) | (1 << second)
                avail = available_mask()
                if len(mis) > best_size:
                    best = list(mis)
                    best_size = len(mis)
                improved = True
                break

        if not improved:
            for vertex in mis:
                penalties[vertex] += 1
            drop = max(mis, key=lambda u: (penalties[u], rng.random()))
            mis.remove(drop)
            mis_mask &= ~(1 << drop)
            avail = available_mask()

    return best


def solve_dense_complement(
    adjacency_list: list[list[int]],
    budget_seconds: float,
    seed: int,
    deadline: float | None = None,
) -> list[int]:
    """
    Find a large maximal clique on dense lambda graphs via complement MIS.

    Max clique in G equals max independent set in the complement of G.
    """
    neighbor_sets = [set(neighbors) for neighbors in adjacency_list]
    n = len(neighbor_sets)
    if n == 0:
        return []

    comp_neighbors = complement_neighbor_sets(neighbor_sets)
    comp_masks = neighbor_sets_to_masks(comp_neighbors) if n <= MAX_BITSET_NODES else None
    penalties = [0] * n
    rng = random.Random(seed)

    start = time.perf_counter()
    if deadline is None:
        deadline = start + budget_seconds
    else:
        deadline = min(deadline, start + budget_seconds)

    heuristic_deadline = start + max(0.08, budget_seconds * 0.35)
    local_deadline = start + max(0.15, budget_seconds * 0.65)

    mis = mis_random_restarts(
        comp_neighbors,
        min(heuristic_deadline, deadline),
        rng,
        comp_masks=comp_masks,
    )
    clique = extend_to_maximal_clique(adjacency_list, sorted(mis))

    if time.perf_counter() < local_deadline:
        mis = mis_local_search(
            comp_neighbors,
            clique,
            min(local_deadline, deadline),
            rng,
        )
        clique = extend_to_maximal_clique(adjacency_list, sorted(mis))

    if comp_masks is not None and time.perf_counter() < deadline:
        mis = mis_bitset_local_search(
            comp_masks,
            n,
            clique,
            deadline,
            rng,
            penalties=penalties,
        )
        clique = extend_to_maximal_clique(adjacency_list, sorted(mis))

    while time.perf_counter() < deadline:
        burst = min(deadline, time.perf_counter() + 0.08)
        mis = mis_random_restarts(comp_neighbors, burst, rng, comp_masks=comp_masks)
        candidate = extend_to_maximal_clique(adjacency_list, sorted(mis))
        burst = min(deadline, time.perf_counter() + 0.12)
        candidate = mis_local_search(comp_neighbors, candidate, burst, rng)
        candidate = extend_to_maximal_clique(adjacency_list, sorted(candidate))
        if len(candidate) > len(clique):
            clique = candidate

    if not is_valid_maximum_clique(adjacency_list, clique):
        seed_vertex = min(range(n), key=lambda v: len(comp_neighbors[v]))
        clique = extend_to_maximal_clique(adjacency_list, [seed_vertex])

    return sorted(clique)

File: crates/model_upgrade_core/src/dense.rs
use std::collections::HashSet;
use std::time::Instant;

use rand::prelude::*;
use rand::rngs::StdRng;

use crate::bitsets::{mask_degree, neighbor_sets_to_masks, BitMask, MAX_BITSET_NODES};
use crate::graph_utils::core_and_degeneracy;
use crate::validation::{extend_to_maximal_clique, is_valid_maximum_clique};

pub const DENSE_DEGREE_RATIO: f64 = 0.70;

pub fn is_dense_graph(neighbor_sets: &[HashSet<usize>], threshold: f64) -> bool {
    let n = neighbor_sets.len();
    if n <= 1 {
        return false;
    }
    let avg_degree: f64 = neighbor_sets.iter().map(|s| s.len()).sum::<usize>() as f64 / n as f64;
    avg_degree / (n - 1) as f64 >= threshold
}

pub fn complement_neighbor_sets(neighbor_sets: &[HashSet<usize>]) -> Vec<HashSet<usize>> {
    let n = neighbor_sets.len();
    let universe: HashSet<usize> = (0..n).collect();
    (0..n)
        .map(|vertex| {
            let mut comp = universe.clone();
            comp.remove(&vertex);
            comp = comp
                .difference(&neighbor_sets[vertex])
                .copied()
                .collect();
            comp
        })
        .collect()
}

pub fn greedy_mis(comp_neighbors: &[HashSet<usize>], vertex_order: &[usize]) -> Vec<usize> {
    let mut mis: Vec<usize> = Vec::new();
    let mut blocked: HashSet<usize> = HashSet::new();
    for &vertex in vertex_order {
        if blocked.contains(&vertex) {
            continue;
        }
        mis.push(vertex);
        blocked.insert(vertex);
        blocked.extend(comp_neighbors[vertex].iter().copied());
    }
    mis
}

pub fn greedy_mis_grasp(
    comp_neighbors: &[HashSet<usize>],
    rng: &mut StdRng,
    start_vertex: Option<usize>,
    rcl_size: usize,
    comp_masks: Option<&[BitMask]>,
) -> Vec<usize> {
    let n = comp_neighbors.len();
    let use_masks = comp_masks.is_some() && n <= MAX_BITSET_NODES;

    let mut mis: Vec<usize> = Vec::new();
    if let Some(sv) = start_vertex {
        mis.push(sv);
    }

    if use_masks {
        let masks = comp_masks.unwrap();
        let mut blocked_mask = BitMask::empty();
        for &vertex in &mis {
            blocked_mask.set(vertex);
            blocked_mask.or_assign(&masks[vertex]);
        }

        loop {
            let mut available_mask = BitMask::all(n);
            available_mask.and_not_assign(&blocked_mask);
            if available_mask.is_empty() {
                break;
            }
            let candidates = available_mask.vertices();
            let mut ranked = candidates.clone();
            ranked.sort_by_key(|&v| mask_degree(masks[v].and(available_mask)));
            let top_len = rcl_size.min(ranked.len());
            let chosen = ranked[rng.random_range(0..top_len)];
            mis.push(chosen);
            blocked_mask.set(chosen);
            blocked_mask.or_assign(&masks[chosen]);
        }
        return mis;
    }

    let mut blocked: HashSet<usize> = HashSet::new();
    for &vertex in &mis {
        blocked.insert(vertex);
        blocked.extend(comp_neighbors[vertex].iter().copied());
    }

    loop {
        let available: HashSet<usize> = (0..n).filter(|v| !blocked.contains(v)).collect();
        if available.is_empty() {
            break;
        }
        let mut ranked: Vec<usize> = available.iter().copied().collect();
        ranked.sort_by_key(|&v| {
            comp_neighbors[v]
                .intersection(&available)
                .count()
        });
        let top_len = rcl_size.min(ranked.len());
        let chosen = ranked[rng.random_range(0..top_len)];
        mis.push(chosen);
        blocked.insert(chosen);
        blocked.extend(comp_neighbors[chosen].iter().copied());
    }
    mis
}

pub fn mis_random_restarts(
    comp_neighbors: &[HashSet<usize>],
    deadline: Instant,
    rng: &mut StdRng,
    comp_masks: Option<&[BitMask]>,
) -> Vec<usize> {
    let n = comp_neighbors.len();
    let comp_degrees: Vec<usize> = comp_neighbors.iter().map(|s| s.len()).collect();
    let mut best: Vec<usize> = Vec::new();

    let (_, degeneracy) = core_and_degeneracy(comp_neighbors);
    let static_orders: Vec<Vec<usize>> = vec![
        {
            let mut o: Vec<usize> = (0..n).collect();
            o.sort_by_key(|&v| comp_degrees[v]);
            o
        },
        {
            let mut o: Vec<usize> = (0..n).collect();
            o.sort_by_key(|&v| std::cmp::Reverse(comp_degrees[v]));
            o
        },
        {
            let mut o: Vec<usize> = (0..n).collect();
            o.sort_by(|&a, &b| (comp_degrees[a], a).cmp(&(comp_degrees[b], b)));
            o
        },
        (0..n).collect(),
        (0..n).rev().collect(),
        degeneracy.clone(),
        degeneracy.iter().rev().copied().collect(),
    ];

    for order in &static_orders {
        if Instant::now() >= deadline {
            break;
        }
        let candidate = greedy_mis(comp_neighbors, order);
        if candidate.len() > best.len() {
            best = candidate;
        }
    }

    let mut sparse_starts: Vec<usize> = (0..n).collect();
    sparse_starts.sort_by_key(|&v| comp_degrees[v]);
    sparse_starts.truncate(24.min(n));

    for &start in &sparse_starts {
        if Instant::now() >= deadline {
            break;
        }
        let candidate = greedy_mis_grasp(comp_neighbors, rng, Some(start), 5, comp_masks);
        if candidate.len() > best.len() {
            best = candidate;
        }
    }

    while Instant::now() < deadline {
        let candidate = if rng.random_bool(0.55) && !sparse_starts.is_empty() {
            let start = sparse_starts[rng.random_range(0..sparse_starts.len())];
            let rcl = rng.random_range(4..=8);
            greedy_mis_grasp(comp_neighbors, rng, Some(start), rcl, comp_masks)
        } else {
            let mut order: Vec<usize> = (0..n).collect();
            order.shuffle(rng);
            greedy_mis(comp_neighbors, &order)
        };
        if candidate.len() > best.len() {
            best = candidate;
        }
    }

    best
}

fn mis_compatible(comp_neighbors: &[HashSet<usize>], mis_set: &HashSet<usize>, vertex: usize) -> bool {
    mis_set.intersection(&comp_neighbors[vertex]).next().is_none()
}

pub fn mis_local_search(
    comp_neighbors: &[HashSet<usize>],
    mis: &[usize],
    deadline: Instant,
    rng: &mut StdRng,
) -> Vec<usize> {
    let mut best: Vec<usize> = mis.to_vec();
    let mut best_set: HashSet<usize> = best.iter().copied().collect();
    let n = comp_neighbors.len();

    while Instant::now() < deadline {
        if best.is_empty() {
            break;
        }

        let remove_vertices: Vec<usize> = if rng.random_bool(0.2) && best.len() >= 2 {
            let mut ranked = best.clone();
            ranked.sort_by_key(|&v| comp_neighbors[v].len());
            ranked[..2].to_vec()
        } else if rng.random_bool(0.75) {
            let v = *best
                .iter()
                .min_by_key(|&&v| comp_neighbors[v].len())
                .unwrap();
            vec![v]
        } else {
            vec![*best.choose(rng).unwrap()]
        };

        let reduced: HashSet<usize> = best_set
            .difference(&remove_vertices.iter().copied().collect())
            .copied()
            .collect();
        let candidates: Vec<usize> = (0..n)
            .filter(|&vertex| {
                !reduced.contains(&vertex) && mis_compatible(comp_neighbors, &reduced, vertex)
            })
            .collect();
        if candidates.is_empty() {
            continue;
        }

        let mut improved = false;

        if remove_vertices.len() == 1 {
            let add_vertex = *candidates
                .iter()
                .min_by_key(|&&v| comp_neighbors[v].len())
                .unwrap();
            let mut trial: Vec<usize> = reduced.iter().copied().collect();
            trial.push(add_vertex);
            trial.sort_unstable();
            if trial.len() >= best.len() {
                best = trial;
                best_set = best.iter().copied().collect();
                improved = true;
            }

            if !improved && best.len() >= 2 && rng.random_bool(0.4) {
                'outer: for (i, &u) in candidates.iter().enumerate() {
                    for &v in &candidates[i + 1..] {
                        if !comp_neighbors[u].contains(&v) {
                            let mut pair_trial: Vec<usize> = reduced.iter().copied().collect();
                            pair_trial.push(u);
                            pair_trial.push(v);
                            pair_trial.sort_unstable();
                            if pair_trial.len() > best.len() {
                                best = pair_trial;
                                best_set = best.iter().copied().collect();
                                improved = true;
                                break 'outer;
                            }
                        }
                    }
                }
            }
        }

        if !improved && remove_vertices.len() == 1 {
            let add_vertex = *candidates.choose(rng).unwrap();
            let mut trial: Vec<usize> = reduced.iter().copied().collect();
            trial.push(add_vertex);
            trial.sort_unstable();
            if trial.len() == best.len() {
                best = trial;
                best_set = best.iter().copied().collect();
            }
        }
    }

    best
}

pub fn mis_bitset_local_search(
    comp_masks: &[BitMask],
    n: usize,
    init_mis: &[usize],
    deadline: Instant,
    rng: &mut StdRng,
    penalties: &mut [i32],
) -> Vec<usize> {
    let mut mis: Vec<usize> = init_mis.to_vec();
    let mut mis_mask = BitMask::empty();
    for &vertex in &mis {
        mis_mask.set(vertex);
    }

    let mut best = mis.clone();
    let mut best_size = mis.len();
    let full_mask = BitMask::all(n);

    let mut avail = {
        let mut blocked = mis_mask;
        for &vertex in &mis {
            blocked.or_assign(&comp_masks[vertex]);
        }
        let mut a = full_mask;
        a.and_not_assign(&blocked);
        a
    };

    let mut check_counter = 0u32;

    loop {
        check_counter += 1;
        if check_counter >= 256 {
            check_counter = 0;
            if Instant::now() >= deadline {
                break;
            }
        }

        if !avail.is_empty() {
            let chosen = if rng.random_bool(0.65) {
                let mut best_vertex = 0usize;
                let mut best_score = i32::MAX;
                let mut first = true;
                for vertex in avail.iter_bits() {
                    let score =
                        (comp_masks[vertex].and(avail).count() as i32) - penalties[vertex];
                    if first || score < best_score {
                        best_score = score;
                        best_vertex = vertex;
                        first = false;
                    }
                }
                best_vertex
            } else {
                let candidates: Vec<usize> = avail.vertices();
                *candidates.choose(rng).unwrap()
            };

            mis.push(chosen);
            mis_mask.set(chosen);
            {
                let mut blocked = mis_mask;
                for &vertex in &mis {
                    blocked.or_assign(&comp_masks[vertex]);
                }
                avail = full_mask;
                avail.and_not_assign(&blocked);
            }

            if mis.len() > best_size {
                best = mis.clone();
                best_size = mis.len();
            }
            continue;
        }

        if mis.is_empty() {
            break;
        }

        let mut improved = false;
        for _ in 0..mis.len().min(4) {
            let pivot = *mis.choose(rng).unwrap();
            let reduced_mask = {
                let mut m = mis_mask;
                m.clear(pivot);
                m
            };
            let mut mask = full_mask;
            mask.and_not_assign(&reduced_mask);
            for &vertex in &mis {
                if vertex != pivot {
                    mask.and_not_assign(&comp_masks[vertex]);
                }
            }
            mask.and_not_assign(&reduced_mask);

            let mut pair: Option<(usize, usize)> = None;
            for first in mask.iter_bits() {
                let mut rest = mask;
                rest.and_not_assign(&comp_masks[first]);
                rest.clear(first);
                if let Some(second) = rest.lowest_bit() {
                    pair = Some((first, second));
                    break;
                }
            }

            if let Some((first, second)) = pair {
                mis.retain(|&v| v != pivot);
                mis_mask.clear(pivot);
                mis.push(first);
                mis.push(second);
                mis_mask.set(first);
                mis_mask.set(second);
                {
                    let mut blocked = mis_mask;
                    for &vertex in &mis {
                        blocked.or_assign(&comp_masks[vertex]);
                    }
                    avail = full_mask;
                    avail.and_not_assign(&blocked);
                }
                if mis.len() > best_size {
                    best = mis.clone();
                    best_size = mis.len();
                }
                improved = true;
                break;
            }
        }

        if !improved {
            for &vertex in &mis {
                penalties[vertex] += 1;
            }
            let drop = *mis
                .iter()
                .max_by(|&&a, &&b| {
                    penalties[a]
                        .cmp(&penalties[b])
                        .then_with(|| rng.random::<u32>().cmp(&rng.random()))
                })
                .unwrap();
            mis.retain(|&v| v != drop);
            mis_mask.clear(drop);
            {
                let mut blocked = mis_mask;
                for &vertex in &mis {
                    blocked.or_assign(&comp_masks[vertex]);
                }
                avail = full_mask;
                avail.and_not_assign(&blocked);
            }
        }
    }

    best
}

pub fn solve_dense_complement(
    adjacency_list: &[Vec<usize>],
    budget_seconds: f64,
    seed: u64,
    deadline: Option<Instant>,
) -> Vec<usize> {
    let neighbor_sets: Vec<HashSet<usize>> = adjacency_list
        .iter()
        .map(|n| n.iter().copied().collect())
        .collect();
    let n = neighbor_sets.len();
    if n == 0 {
        return vec![];
    }

    let comp_neighbors = complement_neighbor_sets(&neighbor_sets);
    let comp_masks = if n <= MAX_BITSET_NODES {
        Some(neighbor_sets_to_masks(
            &comp_neighbors
                .iter()
                .map(|s| {
                    let mut v: Vec<usize> = s.iter().copied().collect();
                    v.sort_unstable();
                    v
                })
                .collect::<Vec<_>>(),
        ))
    } else {
        None
    };
    let mut penalties = vec![0i32; n];
    let mut rng = StdRng::seed_from_u64(seed);

    let start = Instant::now();
    let deadline = match deadline {
        Some(d) => d.min(start + std::time::Duration::from_secs_f64(budget_seconds)),
        None => start + std::time::Duration::from_secs_f64(budget_seconds),
    };

    let heuristic_deadline = start
        + std::time::Duration::from_secs_f64(budget_seconds * 0.35f64.max(0.08));
    let local_deadline = start
        + std::time::Duration::from_secs_f64(budget_seconds * 0.65f64.max(0.15));

    let mis = mis_random_restarts(
        &comp_neighbors,
        heuristic_deadline.min(deadline),
        &mut rng,
        comp_masks.as_deref(),
    );
    let mut clique = extend_to_maximal_clique(adjacency_list, &mis);

    if Instant::now() < local_deadline {
        let mis = mis_local_search(
            &comp_neighbors,
            &clique,
            local_deadline.min(deadline),
            &mut rng,
        );
        clique = extend_to_maximal_clique(adjacency_list, &mis);
    }

    if comp_masks.is_some() && Instant::now() < deadline {
        let mis = mis_bitset_local_search(
            comp_masks.as_ref().unwrap(),
            n,
            &clique,
            deadline,
            &mut rng,
            &mut penalties,
        );
        clique = extend_to_maximal_clique(adjacency_list, &mis);
    }

    while Instant::now() < deadline {
        let burst = deadline.min(Instant::now() + std::time::Duration::from_secs_f64(0.08));
        let mis = mis_random_restarts(&comp_neighbors, burst, &mut rng, comp_masks.as_deref());
        let mut candidate = extend_to_maximal_clique(adjacency_list, &mis);
        let burst = deadline.min(Instant::now() + std::time::Duration::from_secs_f64(0.12));
        let mis = mis_local_search(&comp_neighbors, &candidate, burst, &mut rng);
        candidate = extend_to_maximal_clique(adjacency_list, &mis);
        if candidate.len() > clique.len() {
            clique = candidate;
        }
    }

    if !is_valid_maximum_clique(adjacency_list, &clique) {
        let seed_vertex = (0..n)
            .min_by_key(|&v| comp_neighbors[v].len())
            .unwrap();
        clique = extend_to_maximal_clique(adjacency_list, &[seed_vertex]);
    }

    clique.sort_unstable();
    clique
}
